File: backend/src/domain/entities/ProductListing.ts
export type ProductListingId = string;

export type ProductListingInput = {
  shopifyProductId?: string | null;
  title: string;
  handle?: string | null;
  vendor?: string | null;
  productType?: string | null;
  description?: string | null;
  tags?: string[];
  price?: number | null;
  currency?: string;
  inventoryQuantity?: number | null;
  rawData?: Record<string, unknown>;
};

export type ProductListing = ProductListingInput & {
  id: ProductListingId;
  tags: string[];
  currency: string;
  rawData: Record<string, unknown>;
  createdAt: Date;
  updatedAt: Date;
};

export type ProductListingWithSuggestion = ProductListing & {
  latestSuggestion: OptimizationSuggestionSummary | null;
};

export type OptimizationSuggestionSummary = {
  id: string;
  agentRunId: string;
  improvedTitle: string | null;
  improvedDescription: string;
  tags: string[];
  positioningRecommendations: string | null;
  seoNotes: string | null;
  score: number | null;
  rationale: string | null;
  accepted: boolean;
  createdAt: Date;
};

File: app/models/product_listing.rb
class ProductListing < ApplicationRecord
  has_many :agent_runs, dependent: :destroy
  has_many :optimization_suggestions, dependent: :destroy
  has_one  :latest_suggestion,
    -> { order(created_at: :desc) },
    class_name: "OptimizationSuggestion"

  validates :title, presence: true
  validates :currency, format: { with: /\A[A-Za-z]{3}\z/ }
  validates :price, numericality: { greater_than_or_equal_to: 0, allow_nil: true }

  before_validation :set_defaults

  def as_api_json
    suggestion = latest_suggestion
    {
      id: id,
      shopifyProductId: shopify_product_id,
      title: title,
      handle: handle,
      vendor: vendor,
      productType: product_type,
      description: description,
      tags: tags || [],
      price: price,
      currency: currency,
      inventoryQuantity: inventory_quantity,
      rawData: raw_data || {},
      latestSuggestion: suggestion ? suggestion.as_api_json : nil,
      createdAt: created_at,
      updatedAt: updated_at
    }
  end

  private

  def set_defaults
    self.currency ||= "USD"
    self.tags ||= []
    self.raw_data ||= {}
  end
end
